publishing

mshafer1 opened the pull request:
Initial publishing of DB

TODO: migrate existing tests (currently use concrete implementations of DataBaseItem that are not ready to be public) and include in this depot

TODO: make a nuget

mshafer1 committed:
publishing

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+    {
+        protected const float db_version = 1.2f;
+        protected const float minimum_compatible_version = 1;
+        private static JsonSerializerSettings jsonSerializerSettings = new JsonSerializerSettings()
```
Should this be readonly?

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+            //    mut.ReleaseMutex();
+        }
+
+        public void Dispose()
```
so why exactly do you want this thing to be Disposable? If Dispose doesn't do anything, what's the point? Consider adding a comment when you do something weird, like implement an interface but then leave the implementation blank.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+            // update the DB
+        }
+
+        //public static DataBase<T> mergeDBs(DataBase<T> primary, DataBase<T> secondary, Func<DataBase<T>, DataBase<T>, bool> Conflict, Func<T, T, T> resolve)
```
don't check in commented code unless it's acting as an example.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+        public delegate void TChangedEventHandler(object sender, ID id);
+        public event TChangedEventHandler ItemChanged;
+
+        public T Find(Func<T, bool> select)
```
What benefit does this method provide, other than obfuscation?

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+            return new DataBase<DBTransaction<T>>(transactions_filename, true);
+        }
+
+        private void LoadFile(string file, bool registerItemsForPropertyChange)
```
Some of your methods have capital names, some have lowercase names. I haven't been able to distinguish a pattern. You should make it consistent.

Standard .net code usually has methods capitalized.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+            return count;
+        }
+
+        private int countRecentTransactions(List<TransactionType> transactionTypes, IEnumerable<DBTransaction<T>> list = null)
```
private methods and public methods are interleaved a lot. Consider sorting them to make the code more predictable for the reader.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+        Unknown
+    }
+
+    public class DBTransaction<T> : DatabaseObject where T : DatabaseObject
```
as big as the primary type in this file is, consider moving the auxiliary types to their own files.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+        protected DataBase<DBTransaction<T>> transactions_db { get; set; }
+        #endregion
+
+        #region fields
```
you put your constants at the top and your fields at the bottom... This is a non-standard practice that made me confused as to where the mut and _locker were coming from until I eventually realized that I should just ignore the code and keep scrolling until I got to their definitions.

Generally when organizing code, try to sort it in a way that makes it easiest to read, such as:

* Prerequisites for understand methods (such as constants, fields, and simple properties)
* The main entry points into the type (constructors and public methods)
* Detail oriented private methods with clear names, so that when the reader encounters their usage for the first time there's no question as to their purpose.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+        #region fields
+        // http://www.albahari.com/threading/part2.aspx#_Mutex
+        // Create mutex in constructor - name it so that only one instance of db class can be accessing files
+        protected static Mutex mut;
```
I'm just... so confused. Why is a PROTECTED STATIC **MUTEX** being filled as part of a PUBLIC CONSTRUCTOR? This sends so many mixed messages.

* Every time the public (string) constructor is called, all existing databases lose their mutex and get a new one. The _locker prevents any of them from throwing away a mutex in use, but I can't understand the use of that.
* Databases loaded from Json (or created from subclasses via the protected constructor) inherit the existing mutex, if there is one. This seems like it would cause a lot of bugs and likely needs to be unit tested heavily.
* Subclasses that use the public constructor end up with different behavior than subclasses that use the protected constructor.

Consider having the protected constructor call the public one in cases like this... or just remove the protected constructor all together.

Also, why do you even need both a Mutex and a _locker? The Mutex lets you do things like "try to get a lock but let me know if you fail after X seconds" which is useful, but if you have the Mutex then why do you need the lock? If it's gonna be static anyway, why not just use a static constructor to initialize a readonly Mutex field and not even have a lock?

Ok, and now for the fun one:

statics and generics don't mix in the usual way. That is, every variety of the generic type (for example, if you have two different types of DataBaseObjects) gets its own instance of the static fields. This means that your _locker is NOT fully static, in that it only prevents collisions between two instances of a DataBase that use the same generic parameter.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+        public object property_new { get; set; }
+        [JsonProperty()]
+        public int ID { get; private set; } // switch to using default ID??
+        private static int count = 0;
```
fields and properties intermingled makes this section hard to read.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+    {
+        public DBTransaction() : base()
+        {
+            ID = count++;
```
so is the first used ID end up being 0 or 1?

Trick question: the correct answer is "it's not obvious."

Don't do increment and decrement operations as part of other statements. You're not limited on how many lines you have: make your code as obvious as possible by doing things one at a time. And since 'count' never gets decremented, consider changing its name.

`ID = nextAvailableID;`

`nextAvailableID++;`

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+            property_new = inTransaction.property_new;
+            property_old = inTransaction.property_old;
+
+            count++;
```
so why are we incrementing count here? I mean, if it's actually a transaction count, then sure, you made another transaction. But if it's just to know "what's the next ID I should use" then you didn't use a new ID.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+        public object property_old { get; set; }
+        public object property_new { get; set; }
+        [JsonProperty()]
+        public int ID { get; private set; } // switch to using default ID??
```
consider sorting your properties to make it easier to understand what's going on. For example, you could:

* list all the 'get; private set;' Json Properties first.
* list all the 'get; set;' properties after.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+        private static int count = 0;
+        public bool? Active { get => this.Get(); set => this.Set(value); }
+
+        public static string ITEM_REMOVED { get { return "ITEM_REMOVED"; } }
```
why are these properties instead of constants?

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1039 @@
+        }
+    }
+
+    public class DBException : Exception
```
why is this not the base class of all other DB Exceptions?

haven1433 commented on MiniDB/DataBaseSerializer.cs:
```diff
@@ -0,0 +1,66 @@
+    {
+        // the collection of foo elements
+        public ObservableCollection<T> Collection { get; set; }
+        // the properties of FooCollection to serialize
```
"FooCollection"

haven1433 commented on MiniDB/DataBaseSerializer.cs:
```diff
@@ -0,0 +1,66 @@
+        {
+            // N.B. null handling is missing
+            var surrogate = serializer.Deserialize<DataBaseSurrogate<T>>(reader);
+            var fooElements = surrogate.Collection;
```
foo?

haven1433 commented on MiniDB/DatabaseObject.cs:
```diff
@@ -0,0 +1,28 @@
+    public class DatabaseObject : IDatabaseItem
+    {
+        [JsonProperty()]
+        public virtual ID ID { get; private set; }
```
a public virtual property with a private set doesn't make any sense. You want a subclass to be able to override the get and basically provide their own set that then ignores anything set by the baseclass... yup I'm confused as to how to use that.

Also, since you're only ever calling the setter in the constructor, you can flat out remove it. You might even go one further and remove it from the constructor:

`public ID ID { get; } = new ID();`

haven1433 commented on MiniDB/EncryptedDataBase.cs:
```diff
@@ -0,0 +1,127 @@
+            return "";
+        }
+
+        protected override void _cacheDB()
```
some methods start with a lowercase letter, some start with an underscore, some start with a capital... generally the point of using different starting symbols for different members is to be able to quickly distinguish methods from fields or something like that.

haven1433 commented on MiniDB/EncryptedDataBase.cs:
```diff
@@ -0,0 +1,127 @@
+        {
+            RijndaelManaged RMCrypto = new RijndaelManaged();
+
+            byte[] IV = new byte[Key.Length];
```
IV isn't a very helpful name, I can't tell what this is for without reading the code really close.

haven1433 commented on MiniDB/HardwareID.cs:
```diff
@@ -0,0 +1,57 @@
+        public static byte[] IDValueBytes(string seed)
+        {
+            string id = ID(seed);
+            //Debug.WriteLine(id);
```
I'll just mention again that you should remove commented code. It makes the code harder to read.

mshafer1 committed:
Addressing review comments

Fixed #1 #2 #3 #4 #5 #6 #7 #8 #9 #10 #11 #12 #13 #14 #15 #16

TO-DO: fix naming consistency

TO-DO:  make nuget

mshafer1 committed:
fixing newtonsof nuget on this machine

Installed StyleCop and started making style consistent.

mshafer1 committed:
working on StyleCop 1

mshafer1 committed:
Working on StyleCop 2

mshafer1 committed:
working on style cop 3

mshafer1 committed:
working on style cop 4 (finished with tests passing)

mshafer1 committed:
creating nuget (not yest published)

mshafer1 committed:
setting license, project folder, an releaseNotes on Nuget package

mshafer1 committed:
Updating history comment and adding logo file

mshafer1 commented:
Addressed comments, and created nuget (create with nuget pack in the folder next to the miniDB.csproj file).


TODOne: migrate existing tests (currently use concrete implementations of DataBaseItem that are not ready to be public) and include in this depot

TODOne: make a nuget

TODO: (not blocking publish) make call back for migrating DB versions (current implementation in encryptedDB is not extensible).

TODO: (not blocking publish) prevent a db of dbTransactions from being made (see Database<T> constructor)

TODO: (not blocking publish) Make a compress call back (see Database<T>:SerializeData)

TODO: (not blocking publish) Double check (can we test this??) that the lock object is used everywhere that the db acts on the data

TODO: (not blocking publish) implement DB merging

mshafer1 committed:
adding icon and HardwareID.dll to nuget

mshafer1 committed:
attempting to add Travis CI

mshafer1 committed:
fixing travis project name

mshafer1 committed:
adding xunit runners to travis

mshafer1 committed:
fixing solution name for nugets in travis

mshafer1 committed:
maybe this time . . .

mshafer1 committed:
serializing tests on Travis

mshafer1 committed:
I think this time . . .

mshafer1 committed:
adding support for Unix os's ID's

mshafer1 committed:
requiring linux OS

mshafer1 committed:
requiring linux OS

mshafer1 committed:
Merge branch 'initial-publish' of https://github.com/mshafer1/miniDB into initial-publish

mshafer1 committed:
using different db files to fix threading issue

mshafer1 committed:
travis not working yet . . .

mshafer1 committed:
travis not working yet . . .

mshafer1 committed:
fixing rename

mshafer1 committed:
cleaning our console runner

mshafer1 committed:
trying with destructors

mshafer1 committed:
re-enable travis

mshafer1 committed:
add migration call back and corresponding tests.

mshafer1 committed:
add migration call back and corresponding tests.

mshafer1 committed:
Merge branch 'initial-publish' of https://github.com/mshafer1/miniDB into initial-publish

mshafer1 committed:
disable travis - not pursuing that avenue anymore due to unable to resolve issues

mshafer1 committed:
trying to fix travis file

mshafer1 committed:
disable travis - because of unresovable issues

haven1433 commented on DbXunitTests/DBLoadTests.cs:
```diff
@@ -0,0 +1,200 @@
+        [Fact]
+        public void TestCanUseUsingToReloadSameDBTypeWithSameFile()
+        {
+            Console.WriteLine($"Test Can work in using statement");
```
Generally you shouldn't need to have print statements in your tests. If you need a print statement at the beginning of your test to help you know what's going on during debugging, then you need better names for your tests.

haven1433 commented on DbXunitTests/DBLoadTests.cs:
```diff
@@ -0,0 +1,200 @@
+        /// </summary>
+        private void Cleanup()
+        {
+            if (File.Exists(this.filename))
```
note that if the object destructor runs this, it's possible for filename / filename2 / transactionsFile to be null, or to point to a string that no longer exists. Consider either removing the destructor (because your code never expects it to be needed) or adding a comment explaining why you want/need an explicit destructor on a test. Similarily, you might remove the IDispose implementation all together and just rely on Finally getting called.

haven1433 commented on DbXunitTests/DBLoadTests.cs:
```diff
@@ -0,0 +1,200 @@
+                File.Delete(this.filename2);
+            }
+
+            if (File.Exists(this.transactionsFile))
```
You have the same logic written 3 times. Consider refactoring with a local method or LINQ.

Action<string> delete = str => { if (File.Exists(str) File.Delete(str); }
delete(this.filename);
delete(this.filename2);
delete(this.transactionsFile);

// or

new[] { this.filename, this.filename2, this.transactionsFile }.Where(File.Exists).ToList().ForEach(File.Delete);

haven1433 commented on DbXunitTests/ExampleComplicatedStoredItem.cs:
```diff
@@ -0,0 +1,156 @@
+        /// <summary>
+        /// Gets the full name of the stored person
+        /// </summary>
+        public string Name
```
Consider minifying.

public string Name => $"{FirstName} {LastName}";

haven1433 commented on DbXunitTests/ExampleComplicatedStoredItem.cs:
```diff
@@ -0,0 +1,156 @@
+        #region private methods
+
+        /// <summary>
+        /// When an address field is changed - raise an event on this object to
```
*too

haven1433 commented on DbXunitTests/ExampleComplicatedStoredItem.cs:
```diff
@@ -0,0 +1,156 @@
+        /// </summary>
+        public AddressClass()
+        {
+            this.FirstLine = this.SecondLine = this.City = this.State = this.Zip = string.Empty;
```
I would caution against ever doing this in production code... but I've done stuff like this in personal projects too.

mshafer1 replied:
What's the alternative? I can't find anything online.

haven1433 commented on DbXunitTests/ExampleComplicatedStoredItem.cs:
```diff
@@ -0,0 +1,156 @@
+        /// </summary>
+        /// <param name="sender">the object that raissed the event - expected to be this address</param>
+        /// <param name="e">the event args</param>
+        private void Address_PropertyChangedExtended(object sender, MiniDB.PropertyChangedExtendedEventArgs e)
```
I kinda feel like you can get this to happen automatically with proper coding in the base class. Some kind of special code that checks in your Set(value) method for the object being set is DatabaseObject. However, I noticed that DatabaseObject never actually learns about the Address property, so there would be some refactoring required to get that to work. At the minimum, some version of the forwarding event handler can live in the base class, so that you could do something like

this.Address.PropertyChangeExtended += this.GeneratePropertyChangeForward(nameof(this.Address), () => this.Address);

haven1433 commented on DbXunitTests/ExampleComplicatedStoredItem.cs:
```diff
@@ -0,0 +1,156 @@
+        /// <summary>
+        /// Gets or sets the city
+        /// </summary>
+        public string City
```
You've done such a good job at simplifying this code that it feels like a real shame that you need the get/set duplication at all. However, DependencyProperties (and my own ViewModelBase) get stuck in similar patterns, with the only real solution being some sort of run-time code generation via dynamic languages or Reflect.Emit. Those aren't worth it in almost all cases, so stick with what you have.

haven1433 commented on DbXunitTests/ExampleStoredItem.cs:
```diff
@@ -0,0 +1,96 @@
+    /// <summary>
+    /// An example of an object class that can be stored in a MiniDB. This is used in the unit tests.
+    /// </summary>
+    internal class ExampleStoredItem : MiniDB.DatabaseObject
```
So in all these types, why not just define "using MiniDB" instead of requiring the fulling qualified MiniDB.Whatever?

mshafer1 replied:
I'm trying to be explicit on what is coming from the MiniDB namespace.

haven1433 commented on DbXunitTests/ExampleStoredItem.cs:
```diff
@@ -0,0 +1,96 @@
+        /// </summary>
+        /// <param name="firstName">First name to store</param>
+        /// <param name="lastName">Last name to store</param>
+        public ExampleStoredItem(string firstName, string lastName) : this()
```
This is clear to read for now because the constructors are so short, but is generally a bad pattern. If you're going to have a constructor call another constructor, then you shouldn't have both constructors setting the same properties. Usually the way you do this is by writing the most complex constructor first (the one with the most parameters) and then have all the other constructors forward to that one, filling in default values as needed. Example:

```
public ExampleStoredItem(string firstName, string lastName, Id id) : base(id)
{
   this.FirstName = firstName;
   this.LastName = lastName;
   this.Age = 0;
}

public ExampleStoredItem(ID id) : this(string.Empty, string.Empty, id) {
   // nop
}
```

Since you have two different base constructors you want to call, this probably means fulling implementing 2 constructors and then having any other constructors just forward to those two.

haven1433 commented on DbXunitTests/ExampleStoredItem.cs:
```diff
@@ -0,0 +1,96 @@
+        /// <summary>
+        /// Gets the full name of the stored person
+        /// </summary>
+        public string Name
```
consider minifying

haven1433 commented on DbXunitTests/ExampleStoredItem.cs:
```diff
@@ -0,0 +1,96 @@
+        /// <summary>
+        /// Gets or sets the age of the example person stored in the db
+        /// </summary>
+        public int? Age
```
do you really want this to be nullable? If this is required to be nullable by the framework for some reason and is meant to be an example for users, you should include an explanation somewhere in the file as to why structs aren't allowed.

haven1433 commented on DbXunitTests/UndoRedoTests.cs:
```diff
@@ -0,0 +1,504 @@
+            using (var db = new MiniDB.DataBase<ExampleStoredItem>(this.filename, 1, 1))
+            {
+                Assert.False(File.Exists(this.filename));
+               this.Sleep(.300);
```
add a comment on why we need to sleep in a test, cuz this is hacky

haven1433 commented on DbXunitTests/UndoRedoTests.cs:
```diff
@@ -0,0 +1,504 @@
+                Assert.False(db.CanRedo);
+                Assert.True(File.Exists(this.filename));
+                Assert.True(File.Exists(this.transactionsFile));
+                Assert.Single(db);
```
go ahead and pull this single out after this, because the final assert (db.First().LastName) looks like a typo at first glance.

haven1433 commented on DbXunitTests/UndoRedoTests.cs:
```diff
@@ -0,0 +1,504 @@
+        /// <summary>
+        /// After all the tests have finished, call the dispose method to cleanup
+        /// </summary>
+        private void Finally()
```
ok I'm convinced at this point that you DON'T want your UNIT TESTS writing/reading databases from files. Change your architecture to allow the database to take in some sort of IStorageStrategy, so that the storage strategy can decide whether the database gets written to / read from a file or just some temporary storage stream. This should greatly simplify your tests and remove all these weird sleeps.

haven1433 commented on DbXunitTests/migrationCallBackTests.cs:
```diff
@@ -0,0 +1,153 @@
+        /// </summary>
+        private void Cleanup()
+        {
+            if (System.IO.File.Exists(this.filename))
```
I'm starting to think you want a public static DeleteIfExists method somewhere.

haven1433 commented on DbXunitTests/migrationCallBackTests.cs:
```diff
@@ -0,0 +1,153 @@
+        /// <summary>
+        /// the expected base id value.
+        /// </summary>
+        private const int ExpectedID = 1955653895;
```
these IDs are rather specific. Consider adding some sort of comment.

haven1433 commented on HardwareID/HardwareID.cs:
```diff
@@ -0,0 +1,256 @@
+    /// </summary>
+    public sealed class HardwareID
+    {
+        #region properties
```
except that these are constants, not properties...

haven1433 commented on HardwareID/HardwareID.cs:
```diff
@@ -0,0 +1,256 @@
+        /// <summary>
+        /// Cache the fields, for quick recapture
+        /// </summary>
+        private static Dictionary<string, object> data = new Dictionary<string, object>();
```
Make the field readonly to lower your chances of shooting yourself in the foot.

haven1433 commented on HardwareID/HardwareID.cs:
```diff
@@ -0,0 +1,256 @@
+        /// <summary>
+        /// Keys to parse from SystemInfo
+        /// </summary>
+        private static Dictionary<string, string> systemInfoFields = new Dictionary<string, string>()
```
This should probably be exposed (or even implemented) as a ReadOnlyDictionary. And you likely want the field to be readonly as well.

haven1433 commented on HardwareID/HardwareID.cs:
```diff
@@ -0,0 +1,256 @@
+        /// <returns>string hash of _data + seed</returns>
+        public static string UniqueID(string seed)
+        {
+            return GetHash($"{_data}\nSeed >> {seed}");
```
use Environment.NewLine for portability.

haven1433 commented on HardwareID/HardwareID.cs:
```diff
@@ -0,0 +1,256 @@
+            byte[] bt = Encoding.ASCII.GetBytes(s);
+
+            // Grab the Hexadecimal value of the MD5 hash
+            return GetHexString(sec.ComputeHash(bt));
```
add a comment explaining why string.GetHashCode() wasn't good enough for your purposes.

haven1433 commented on HardwareID/HardwareID.cs:
```diff
@@ -0,0 +1,256 @@
+        /// <returns>hexadecimal string representation of list</returns>
+        private static string GetHexString(IList<byte> bt)
+        {
+            string s = string.Empty;
```
you want to use StringBuilder here

haven1433 commented on HardwareID/HardwareID.cs:
```diff
@@ -0,0 +1,256 @@
+        /// </summary>
+        /// <param name="name">The name of the item to fetch (default: caller)</param>
+        /// <returns>the chached value</returns>
+        private static dynamic Get([CallerMemberName]string name = null)
```
no reason to get fancy, just return a string.

haven1433 commented on HardwareID/HardwareID.cs:
```diff
@@ -0,0 +1,256 @@
+        /// <param name="value">The desired value to store in data</param>
+        /// <param name="name">The name to store the value under (default: caller)</param>
+        /// <returns>bool - successful or not</returns>
+        private bool Set<T>(T value, [CallerMemberName]string name = null)
```
this method is unused. Remove it, then you can make the class static.

haven1433 commented on MiniDB/DBTransaction.cs:
```diff
@@ -0,0 +1,145 @@
+            this.Property_old = inTransaction.Property_old;
+            
+            // needed? - keep until collision detecting is implemented
+            nextAvailableID++;
```
yeah this line seems silly and not required.

haven1433 commented on MiniDB/DBTransaction.cs:
```diff
@@ -0,0 +1,145 @@
+
+        #region transacted object properties
+        /// <summary>
+        /// Gets or sets the ID of the item that was acted on
```
these code comments are useless. Either make them say "why" instead of "what" or just delete them. And if they're there because Visual Studio wants them there, then you might as well remove them, because then you at least get a warning about having no documentation instead of the false security of bad documentation.

haven1433 commented on MiniDB/DBTransaction.cs:
```diff
@@ -0,0 +1,145 @@
+        /// <summary>
+        /// Gets or sets the type of transaction (Add, delete, modify, etc.) of type <see cref="TransactionType"/>.
+        /// </summary>
+        public TransactionType TransactionType { get => this.Get(); set => this.Set(value); }
```
You probably want some kind of comment explaining why some of these properties are using the Get() and Set() while others don't need to.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1170 @@
+                string global_lock_mutex_name = @"Global\" + mutex_name;
+
+                // from https://stackoverflow.com/a/3111740
+                // try to get existing mutex from syste
```
from syste?

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1170 @@
+                    this.mut = new System.Threading.Mutex(false, global_lock_mutex_name);
+                }
+
+                // acquire the lock from the mutex - this is release in dispose
```
verb tense?

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1170 @@
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the collection currently can redo an undone change
```
this comment is not helpful. Comments should say why, not what.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1170 @@
+            get
+            {
+                // TODOne: this should be number of immediate redo's is less than number of next immediate undo's
+                //  bool result = true;
```
remove commented code

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1170 @@
+        /// Undo last undoable transaction and raise property changed events on changed item.
+        /// TODO: make sure to use locker everywhere that it is messing with the data
+        /// </summary>
+        public void Undo()
```
The Undo() and Redo() methods are far too long. Refactor them into shorter methods. For example, every time you feel the need to add a comment saying what the code is doing, that comment should probably be a method name.

Comments should say why you're doing something, not what you're doing. If you need a comment to explain what you're doing, then your code isn't clear enough yet.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1170 @@
+        /// </summary>
+        /// <param name="last_transaction">what happened that we're changing back</param>
+        /// <param name="transactedItem">the item to act on</param>
+        private static void SetProperty(DBTransaction<T> last_transaction, T transactedItem)
```
Another very long method. A good rule of thumb is that a method is too long to fit on a single screen, then it likely should be refactored into smaller pieces.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1170 @@
+        /// </summary>
+        /// <param name="sender">database that changed (should be this)</param>
+        /// <param name="e">How the collection changed</param>
+        private void DataBase_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
```
shorten / refactor

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1170 @@
+            /// <summary>
+            /// The version of DB that is loaded
+            /// </summary>
+            public float OldVersion;
```
Consider making these get-only and adding a constructor to set them, to better indicate that you don't expect these to change after construction.

haven1433 commented on MiniDB/DatabaseObject.cs:
```diff
@@ -0,0 +1,148 @@
+        /// <summary>
+        /// Event raised when any property is changed (does not contain how changed).
+        /// </summary>
+        public event PropertyChangedEventHandler PropertyChanged;
```
Having this makes me wonder if you should implement INotifyPropertyChanged

haven1433 commented on MiniDB/ID.cs:
```diff
@@ -0,0 +1,241 @@
+        /// <summary>
+        /// A radom number generator to create the next ID number from
+        /// </summary>
+        private static Random random = new Random();
```
readonly?

haven1433 commented on MiniDB/ID.cs:
```diff
@@ -0,0 +1,241 @@
+        public ID Set()
+        {
+            this.hardwareComponent = DBHardwareID.IDValueInt();
+            this.id = random.Next(0, int.MaxValue);
```
and ID can't just be a counter because we are trying to prevent collisions with data in databases that we don't know about? Probably worth a comment somewhere.

haven1433 commented on MiniDB/PropertyChangedExtendedEventArgs.cs:
```diff
@@ -0,0 +1,50 @@
+        /// <param name="oldValue">The old value</param>
+        /// <param name="newValue">The new value</param>
+        /// <param name="undoable">Whether this change is undoable or not</param>
+        public PropertyChangedExtendedEventArgs(string propertyName, object oldValue, object newValue, bool undoable = true)
```
given places where you construct a new PropertyChangedExtenendEventArgs object based on an existing one, you may include a second constructor that takes in a string and a PropertyChangedExtendedEventArgs.

haven1433 commented on MiniDB/DataBase.cs:
```diff
@@ -0,0 +1,1170 @@
+    /// Database Template: template class for persisten observable collection with undo/redo
+    /// </summary>
+    /// <typeparam name="T">The class type to create an observable collection of (must by a child class of DatabaseObject)</typeparam>
+    public class DataBase<T> : ObservableCollection<T>, IDisposable where T : DatabaseObject
```
It's late, so I don't have time to properly review this behemoth. So here's the highlights:

- It looks like your Encryption subclass could be redesigned as a database save strategy, which could go a good distance in simplifying logic in this class (no worrying about subclasses) and that class (no worrying about base classes).

- _Any_ time you need a destructor, add a comment to your code explaining what horrible things can go wrong if the destructor isn't called, and add a Debug.Assert to the destructor to check that it got called in the way you expected.

mshafer1 committed:
some test cleaning

mshafer1 committed:
some cleanup

mshafer1 committed:
working on cleaning - will need to implement strategy still.

mshafer1 committed:
cleaning

mshafer1 committed:
creating (not yest working)

mshafer1 committed:
working on getting json to reload - hung up on transaction db storage strategy.

mshafer1 committed:
trying an interface

mshafer1 committed:
working on setting up

mshafer1 committed:
working on getting it to compile

mshafer1 committed:
working on getting it to compile

mshafer1 committed:
whiped DB and tests

mshafer1 committed:
builds

mshafer1 committed:
working on getting json db up and running

mshafer1 committed:
brought in most of the DataBase class and got it to build.

mshafer1 committed:
working on getting json storage strategy/tests rebuilt.

mshafer1 committed:
Got reload tests working again. Still need to re-implement undo/redo from scratch.

mshafer1 committed:
adding a testing strategy that will help keep track of whether the db is calling the storageStrategy correctly.

mshafer1 committed:
adding transaction types

mshafer1 committed:
getting setup to store transactions.

mshafer1 committed:
got reload with transactions working

mshafer1 committed:
renaming IDatabaseObject to IDBObject for better consistency.

mshafer1 committed:
changing null writer strategy to raise events for checking what is stored.

mshafer1 committed:
tested transactions store and reload

mshafer1 committed:
noting discomfort

mshafer1 committed:
made .Clear() remove each element so that all elements are logged as deleted (TODO: once a transaction manager merges based on time, this should become one transaction)

mshafer1 committed:
made a replace log as an add then delete (plus test)

mshafer1 committed:
working on undo/redo

Added Stub from AutomImplement

mshafer1 committed:
ignore AutoImplement binaries

mshafer1 committed:
due to issue with reference dll, removing code gen from build steps and checking in modified source for stub

mshafer1 committed:
got tests around DB calling undo/redo manager.

mshafer1 committed:
working on setting up undo/redo manager and tests

mshafer1 committed:
got tests building again

Updated AutoImplement, re-instigated build time implement of IUndoRedoManager

mshafer1 committed:
got tests around undo/redo calling storage strategy working again

mshafer1 committed:
working on tests

mshafer1 committed:
making debug mean to run tests

mshafer1 committed:
got basic test of add working

mshafer1 committed:
got undo-redo add test running

Need to implement revert of redo

mshafer1 committed:
got redo/undo stack on add working and tested

mshafer1 committed:
adding test to double check stack in concluded

mshafer1 committed:
working on ModifyTransaction undo - need to pull in reflection from last branch now

mshafer1 committed:
got edit undo-redo working

mshafer1 committed:
got edits of simple types able to undo

mshafer1 committed:
style cop cleanings

mshafer1 committed:
some cleaning

mshafer1 committed:
move owning the transactions cache into the undoRedoManager

mshafer1 committed:
cleaning up comments

mshafer1 committed:
made database automatically register sub items (field or property) that are also IDBObject

mshafer1 committed:
little bit of cleaning

mshafer1 committed:
trying to add Travis again

mshafer1 committed:
trying to trigger travis

mshafer1 committed:
using "copy" instead of "Move" to work be OS agnostic

mshafer1 committed:
adding linux style batch scripts to try to make this work better

mshafer1 committed:
move prebuild steps to manual step for Travis to not run

mshafer1 committed:
check if the other tests will pass on Travis

mshafer1 committed:
stripping down to causing test

mshafer1 committed:
stripping down to find problem test

mshafer1 committed:
removing the broken one

mshafer1 committed:
Revert "stripping down to find problem test"

This reverts commit 28794b0d0e52cee56eb44b856f4022c6c32123ea.

mshafer1 committed:
Revert "stripping down to causing test"

This reverts commit 3a4ea2d207c3eac637f049bb6c96058690a77248.

mshafer1 committed:
re-comment out tests that fail on Travis

mshafer1 committed:
remove tests again to get to passing set

mshafer1 committed:
Revert "remove tests again to get to passing set"

This reverts commit 4364df210c36e94b2177d271720e5a99504d04e2.

mshafer1 committed:
still trying to find a base passing set

mshafer1 committed:
Revert "still trying to find a base passing set"

This reverts commit 1da3743ba4544cb88bf1452730cc4f9ec64cbbeb.

mshafer1 committed:
trying a different small set

mshafer1 committed:
trying to get mono to handle dispose method correctly

mshafer1 committed:
still trying

mshafer1 committed:
clear reference after disposing

mshafer1 committed:
cleaning the dispose method

mshafer1 committed:
making a simpler cleanup

mshafer1 committed:
testing the framework

mshafer1 committed:
still trying to find where the issue is

mshafer1 committed:
tracking down dispose issue

mshafer1 committed:
strill trying to track down issue

mshafer1 committed:
this might fix it

mshafer1 committed:
try this  crazy idea

mshafer1 committed:
Revert "try this  crazy idea"

This reverts commit e7caeef92578d398aac5a7c635b82cd1d27e9631.

mshafer1 committed:
adding xunit config

mshafer1 committed:
Revert "trying a different small set"

This reverts commit f4fdcd3febad8b24a591685560f4e3f34f71b157.

mshafer1 committed:
trying to build with just dotnet

mshafer1 committed:
switching dotnet version

mshafer1 committed:
trying new target version

mshafer1 committed:
Revert "re-comment out tests that fail on Travis"

This reverts commit 8e6bd799fc97d83ea830156ba3818097ba90c636.

mshafer1 committed:
Revert "check if the other tests will pass on Travis"

This reverts commit 2d6e58a8123cd12040401c0f139be37605f73e4b.

mshafer1 committed:
Revert "removing the broken one"

This reverts commit 1c1a726fd8cf5b2746dc42bf5bf26ad5faf55922.

mshafer1 committed:
cleaning test disable

mshafer1 committed:
switch back to mono

mshafer1 committed:
Switch to fileMutexes

mshafer1 committed:
Add stylecop to build of main project

mshafer1 committed:
added stylecop to tests and got it building

mshafer1 committed:
updated HardwareID

mshafer1 committed:
applied stylecop to MutexLocks

mshafer1 committed:
added stylecop to TestFileLocker

mshafer1 committed:
little more cleaning

mshafer1 commented:
[x] migrate existing tests (currently use concrete implementations of DataBaseItem that are not ready to be public) and include in this depot

[ ]: make a nuget

[ ]: (not blocking publish) make call back for migrating DB versions (current implementation in encryptedDB is not extensible).

[ ]: (not blocking publish) Make a compress call back (see Database:SerializeData)

[ ]: (not blocking publish) Double check (can we test this??) that the lock object is used everywhere that the db acts on the data

[ ]: (not blocking publish) implement DB merging

mshafer1 committed:
use quotes for default parameters

mshafer1 commented:
Interestingly - I think this is actually an indicator of the Mutex locks working correctly - as it appears to have run both builds on docker images on the same host (based on boot time, etc.).

mshafer1 committed:
re-add encrypted storage strategy

mshafer1 committed:
re-enable testing EncryptedDB

mshafer1 commented:
hmmm . . . this runs locally just fine - I'm trying to setup a Travis Docker image to reproduce

mshafer1 committed:
fix path joining for transactions file

mshafer1 commented:
AHA! Two builds passed because I fixed the issue and the third failed due to concurrency with the system lock! It works!

mshafer1 committed:
changing to metadata class rather than access to internals

mshafer1 committed:
move lock file next to resource

mshafer1 committed:
adding test project

mshafer1 committed:
move TestProject to own folder

mshafer1 committed:
get build working

mshafer1 committed:
fix reference for actual build flow

mshafer1 committed:
cleanup newline

mshafer1 committed:
setting up for auto deploy

mshafer1 committed:
bump version number after push

mshafer1 commented:
[x] migrate existing tests (currently use concrete implementations of DataBaseItem that are not ready to be public) and include in this depot

[x]: make a nuget

[ ]: (not blocking publish) make call back for migrating DB versions (current implementation in encryptedDB is not extensible).

[ ]: (not blocking publish) Make a compress call back (see Database:SerializeData)

[ ]: (not blocking publish) Double check (can we test this??) that the lock object is used everywhere that the db acts on the data

[ ]: (not blocking publish) implement DB merging

[ ]: (not blocking publish) figure out a way to test nupkg once built

[x]: (not blocking publish) move FileMutex to use local file

[x]: Get mutex lock tests running in Travis

mshafer1 committed:
fixing usage reference

mshafer1 committed:
Merge branch 'initial-publish' of https://github.com/mshafer1/miniDB into initial-publish

mshafer1 committed:
wokring on adding nuget test project

mshafer1 committed:
Merge branch 'initial-publish' of https://github.com/mshafer1/miniDB into initial-publish

haven1433 commented on DbXunitTests/DBCallsStorageStrategyTests.cs:
```diff
@@ -0,0 +1,107 @@
+
+        public void Dispose()
+        {
+            this.testDB?.Dispose();
```
I find this use of ? confusing. The testDB is readonly, and you assign it a value in the only constructor. In what cases are you expecting `testDB` to be null?

haven1433 commented on DbXunitTests/DBCallsStorageStrategyTests.cs:
```diff
@@ -0,0 +1,107 @@
+        public void DoesWriteTransactionsFileOnInsert()
+        {
+            this.nullWritingStorageStrategy.ClearWroteFlags();
+            Assert.False(this.nullWritingStorageStrategy.WroteTransactionsFlag);
```
this line... seems like it's testing the logic of the nullWritingStorageStrategy. If we don't trust the test-double, then make a test class to hold tests where we verify the behavior of the test double. Otherwise, let's not clutter up the reading of this unrelated test.

haven1433 commented on DbXunitTests/DBCallsStorageStrategyTests.cs:
```diff
@@ -0,0 +1,107 @@
+        [Fact]
+        public void DoesWriteMainFileOnInsert()
+        {
+            Assert.False(this.nullWritingStorageStrategy.WroteFlag);
```
This assert already exists in a test all by itself. We can safely remove it, which also makes the test easier to read.

haven1433 commented on DbXunitTests/DBCallsStorageStrategyTests.cs:
```diff
@@ -0,0 +1,107 @@
+            Assert.False(this.nullWritingStorageStrategy.WroteFlag);
+
+            this.testDB.Remove(item);
+            Assert.True(this.nullWritingStorageStrategy.WroteFlag);
```
This test is a bit murky when I'm trying to mentally divide it into Arrange-Act-Assert. I think what it is should be:

* Arrange: add an item to the testDB and clean the test-double's flags.
* Act: remove an item
* Assert: verify that the storage strategy did the right thing.

The assert in the middle confuses this, because I'd expect us to generally trust the test-double in this test. If we wanted to verify the behavior of the test-double, that should be done elsewhere.

Another possible refactor to think about: consider adding a `Reset` method to the nullWritingStorageStrategy that just clears all the flags. That would make it very apparent in any test that uses it that everything you did to the database up to this point was just setup and you don't care about the operations performed by the strategy before this point.

haven1433 commented on DbXunitTests/DBCallsStorageStrategyTests.cs:
```diff
@@ -0,0 +1,107 @@
+            var item = new ExampleStoredItem();
+            this.testDB.Add(item);
+            this.nullWritingStorageStrategy.ClearWroteFlags();
+            Assert.False(this.nullWritingStorageStrategy.WroteTransactionsFlag);
```
same comment on testing the test-double. You get the idea.

haven1433 commented on DbXunitTests/DBCallsStorageStrategyTests.cs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        [Fact]
+        public void DoesWriteMainFileOnItemEdit()
```
slick behavior ;)

haven1433 commented on DbXunitTests/DBLoadTests.cs:
```diff
@@ -0,0 +1,193 @@
+        /// <summary>
+        /// After all tests have run, clear resources
+        /// </summary>
+        private void Finally()
```
can I get a link to some documentation on this? I don't see anywhere in XUnit that talks about this use of Finally methods, and I don't see anywhere where you're calling it either.

haven1433 commented on DbXunitTests/ExampleComplicatedStoredItem.cs:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    /// <summary>
+    /// Example of an supported nested property
```
"an supported"

haven1433 commented on DbXunitTests/GeneratedCode/StubUndoRedoManager.cs:
```diff
@@ -0,0 +1,58 @@
+using System;
+using System.Collections.Generic;
+
+// this file was created by AutoImplement
```
ooooooo

haven1433 commented on prebuild.bat:
```diff
@@ -0,0 +1,8 @@
+
+%~dp0/packages/HavenSoft.AutoImplement.1.1.1/AutoImplement ./MiniDB.dll IUndoRedoManager
+
+move /Y StubUndoRedoManager.cs  %~dp0/DbXunitTests/GeneratedCode
```
since you have this process as part of your build to generate AutoImplement code, it's a little silly to also check in the AutoImplement code. Consider blacklisting the 'GeneratedCode' folder in your .gitconfig.

mshafer1 replied:
this is manually run when needed - wanted a little more control of exactly how often it happens (then just on successful build of the core library or on before any build of the tests).

haven1433 commented on DbXunitTests/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,38 @@
+[assembly: AssemblyTitle("DbXunitTests")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
```
consider deleting the attributes you're not using or giving them reasonable values.

haven1433 commented on DbXunitTests/SystemTests/UndoRedoWithJsonStorageStrategy.cs:
```diff
@@ -0,0 +1,80 @@
+        public void Dispose()
+        {
+            // clean up when finished
+            this.Cleanup();
```
The fact that some of your tests still need access to the external filesystem is a bit of a code smell. I allowed it for my tests only when I wanted to test against real artifacts that actually were expected to exist on disk.

These tests care that you have a source of JSON, and it makes sense for that source to by a stream. But does it have to be a file stream? With a little refactoring, you could probably abstract away the filesystem into a stream manager. The real one could use file streams, but there's no reason why you couldn't have a separate version that uses MemoryStreams.

Using files here is probably ok for a personal project, and XUnit guarantees that tests within the same collection (or class) won't run in parallel, so you don't have to worry about your tests biting eachother. But the point is that you *could* switch completely away from the filesystem if you wanted to, and it would possibly speed up your tests.

mshafer1 replied:
these are sanity tests of the json storage strategy.

haven1433 commented on DbXunitTests/UndoRedoTests/DBMassEditStoresTransactionsTests.cs:
```diff
@@ -0,0 +1,174 @@
+        private void Cleanup()
+        {
+            // if using NullWriter, no files to clean up, but clear the DB
+            this.testDB?.Dispose();
```
pretty sure there's no way this will ever be null

haven1433 commented on HardwareID/HardwareID.cs:
```diff
@@ -0,0 +1,231 @@
+        /// </summary>
+        /// <param name="bt">Byte list</param>
+        /// <returns>hexadecimal string representation of list</returns>
+        private static string GetHexString(IList<byte> bt)
```
since you're adding dashes half the time, consider including an example in the comment header.

haven1433 commented on HardwareID/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("HardwareID")]
+[assembly: AssemblyDescription("")]
```
consider removing any attributes that you're not using

haven1433 commented on MiniDB/Interfaces/IDBObject.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace MiniDB.Interfaces
+{
+    public interface IDBObject
```
a bit of a downer that this doesn't extend from INotifyPropertyChanged. Seems like it would be useful to be able to work things expecting that interface, especially since we use the same event handler.

haven1433 commented on MiniDB/Transactions/UndoRedoManager.cs:
```diff
@@ -0,0 +1,372 @@
+
+        public bool CheckCanUndo()
+        {
+            return this.transactions_DB.Count() > 0 && this.transactions_DB.Count() >
```
consider making a variable for the number of undo transactions, to make this read easier.

    var undoTransactions = this.transactions_DB.Count(transaction => transaction.DBTransactionType == DBTransactionType.Undo)
    var count = this.transactions_DB.Count();
    return count > 0 && count > undoTransactions * 2;

haven1433 commented on MiniDB/Transactions/UndoRedoManager.cs:
```diff
@@ -0,0 +1,372 @@
+                    (2 * this.transactions_DB.Count(x => x.DBTransactionType == DBTransactionType.Undo));
+        }
+
+        public bool CheckCanRedo()
```
so if I undo, undo, undo, redo, undo, redo, I should still be able to redo, right? But if I understand the logic right, my 'redos_count' is currently 0, and my 'undos_count' will be 0, because the 3rd-to-last item is a redo, not an undo.

I think what you want to do is count the number of recent undos minus the number of recent redos, and then return true if the difference is more than zero, counting all recent transactions until you hit something that isn't an undo or redo. This could use some unit testing to expose and improve.

mshafer1 replied:
adding `Test_LongChaneOfChanges_CanUndo` and `Test_LongChaneOfChanges_UndoYieldsCorrectResult` to `DBDefaultUndoRedoManagerTests.cs` to show how this should behave.

haven1433 commented on MiniDB/Transactions/UndoRedoManager.cs:
```diff
@@ -0,0 +1,372 @@
+
+        public void InsertTransaction(IDBTransaction transaction)
+        {
+            this.transactions_DB.Insert(0, transaction);
```
so inserting and removing from an ObservableCollection in an O(1) operation most of the time... if you're adding/removing from the end of the collection. From the start, it's an O(N) operation. Might not matter for small lists, but it'll give you performance trouble for long-lived databases. Consider flipping the internal state around such that new transactions get added to the *end* of the collection.

mshafer1 replied:
I'll have to do some testing, whether it is worse to insert at 0 for large collections, or call reverse to do a count from the end.

haven1433 commented on MiniDB/Transactions/UndoRedoManager.cs:
```diff
@@ -0,0 +1,372 @@
+        /// </summary>
+        /// <param name="last_transaction">what happened that we're changing back</param>
+        /// <param name="transactedItem">the item to act on</param>
+        private static void SetProperty(ModifyTransaction last_transaction, IDBObject transactedItem)
```
this is a very long function. Consider refactoring into smaller pieces.

haven1433 commented on MiniDB/Transactions/UndoRedoManager.cs:
```diff
@@ -0,0 +1,372 @@
+                    Type t = lastObject.GetType();
+                    if (!t.IsGenericType || t.GetGenericTypeDefinition() != typeof(Dictionary<,>))
+                    {
+                        throw new DBCannotUndoException($"Property {lastObject} is not a dictionary, but was used with indexers");
```
fun fact: you can use indexers from lots of base classes. For example, List<T>. But more importantly, you can put indexers an any type by implementing an indexer property. As you noticed, this is exposed through an the `Item` property. You can check AutoImplement to see how I'm dealing with it there.

At the minimum, you should update the error to something like "indexer properties on non-dictionary types are not supported by MiniDB."

haven1433 commented on MiniDB/Transactions/UndoRedoManager.cs:
```diff
@@ -0,0 +1,372 @@
+        /// <param name="transactionType">transaction type to count</param>
+        /// <param name="list">(optional) the list to search (defaults to this)</param>
+        /// <returns>the count of matches</returns>
+        private int CountRecentTransactions(DBTransactionType transactionType, IEnumerable<IDBTransaction> list)
```
this (and the next method) look like they're meant to be static

haven1433 commented on MiniDB/Transactions/UndoRedoManager.cs:
```diff
@@ -0,0 +1,372 @@
+                    temp_list = temp_list.Skip(1);
+                }
+            }
+            while (last_transaction == null || last_transaction.DBTransactionType == notTransactionType);
```
note that this condition is true when you start the loop. So you could switch to using a while loop if you wanted.

I personally find while loops easier to read than do-while loops.

haven1433 commented on MutexLocks/FileMutex.cs:
```diff
@@ -0,0 +1,50 @@
+            }
+            catch (IOException)
+            {
+                throw new MutexException($"Cannot open file {this.file_name}, another process is using it.");
```
I feel like I'm missing something here. But in my mind, Mutex.Get() should either return false if it fails, or block until the resource is available. Having Mutex.Get() throw seems like unexpected behavior.

haven1433 commented on MutexLocks/MutexObject.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace MutexLocks
+{
+    public class MutexObject : IDisposable
```
This class feels rather pointless to me. All you've done is decorated an IDisposable with another class that adds an alias to the one method that IDisposables already have. I was expecting to see some mutex locking behavior in here.

haven1433 reviewed (commented):
Definitely better than the first time I reviewed it. But it looks like some of my feedback from the first review still hasn't been resolved.
